File: generate_bang_gia.py
import random
import threading
import time
import json
import sys
from datetime import datetime
from pathlib import Path

try:
    from vndirect_realtime import create_client, stock_topic, transaction_topic
except ModuleNotFoundError as exc:
    missing_module = exc.name or "dependency"
    raise SystemExit(
        "Thieu thu vien de chay realtime feed. "
        f"Module bi thieu: {missing_module}. "
        "Hay cai dependency bang lenh: python -m pip install -r requirements.txt"
    ) from exc


BASE_DIR = Path(__file__).resolve().parent
STOCK_ID_DIR = BASE_DIR / "StockIDs"
OUTPUT_FILE = BASE_DIR / "bang_gia_chung_khoan.txt"
RANDOM_COUNT = 5
WAIT_TIMEOUT_SECONDS = 20


def load_stock_codes():
    codes = []
    for path in sorted(STOCK_ID_DIR.glob("*.txt")):
        for line in path.read_text(encoding="utf-8").splitlines():
            code = line.strip().upper()
            if code:
                codes.append(code)
    unique_codes = sorted(set(codes))
    if len(unique_codes) < RANDOM_COUNT:
        raise ValueError(f"Khong du ma co phieu de chon {RANDOM_COUNT} ma ngau nhien.")
    return unique_codes


def select_codes(all_codes, args):
    if not args:
        return random.sample(all_codes, RANDOM_COUNT)

    selected_codes = [code.strip().upper() for code in args if code.strip()]
    if len(selected_codes) != RANDOM_COUNT:
        raise SystemExit(f"Hay truyen dung {RANDOM_COUNT} ma co phieu.")

    valid_codes = set(all_codes)
    invalid_codes = [code for code in selected_codes if code not in valid_codes]
    if invalid_codes:
        raise SystemExit(f"Ma khong hop le: {', '.join(invalid_codes)}")

    if len(set(selected_codes)) != RANDOM_COUNT:
        raise SystemExit("5 ma co phieu phai khac nhau.")

    return selected_codes


def pick_value(*values, default="N/A"):
    for value in values:
        if value not in (None, ""):
            return value
    return default


def format_block(index, record):
    lines = [f"Co phieu {index}"]
    for key, value in record.items():
        lines.append(f"{key}: {value}")
    return "\n".join(lines)


def merge_stock_data(code, sp_data, ba_data, fetched_at):
    sp_data = sp_data or {}
    ba_data = ba_data or {}
    return {
        "Mã chứng khoán": code,
        "Giá tham chiếu": pick_value(sp_data.get("basicPrice")),
        "Giá trần": pick_value(sp_data.get("ceilingPrice")),
        "Giá sàn": pick_value(sp_data.get("floorPrice")),
        "Giá mua 1": pick_value(ba_data.get("bidPrice01")),
        "Khối lượng mua 1": pick_value(ba_data.get("bidQtty01")),
        "Giá mua 2": pick_value(ba_data.get("bidPrice02")),
        "Khối lượng mua 2": pick_value(ba_data.get("bidQtty02")),
        "Giá mua 3": pick_value(ba_data.get("bidPrice03")),
        "Khối lượng mua 3": pick_value(ba_data.get("bidQtty03")),
        "Giá bán 1": pick_value(ba_data.get("offerPrice01")),
        "Khối lượng bán 1": pick_value(ba_data.get("offerQtty01")),
        "Giá bán 2": pick_value(ba_data.get("offerPrice02")),
        "Khối lượng bán 2": pick_value(ba_data.get("offerQtty02")),
        "Giá bán 3": pick_value(ba_data.get("offerPrice03")),
        "Khối lượng bán 3": pick_value(ba_data.get("offerQtty03")),
        "Giá khớp lệnh": pick_value(ba_data.get("matchPrice"), sp_data.get("currentPrice")),
        "Khối lượng khớp lệnh": pick_value(ba_data.get("matchQtty"), sp_data.get("currentQtty")),
        "Thời gian lấy dữ liệu": pick_value(ba_data.get("time"), sp_data.get("time"), fetched_at),
    }


def collect_realtime_rows(selected_codes):
    sp_map = {}
    ba_map = {}
    done = threading.Event()
    topics = [stock_topic(code) for code in selected_codes]
    topics += [transaction_topic(code) for code in selected_codes]

    def on_message(client, userdata, msg):
        try:
            payload = msg.payload.decode("utf-8", errors="replace")
            parsed = json.loads(payload)
        except Exception:
            return

        topic = msg.topic
        parts = topic.split("/")
        if len(parts) < 2:
            return

        if parts[0] == "T" and len(parts) >= 3:
            code = parts[2]
            if isinstance(parsed, dict):
                ba_map[code] = parsed
        else:
            code = parts[1]
            if isinstance(parsed, dict):
                sp_map[code] = parsed

        if all(code in sp_map and code in ba_map for code in selected_codes):
            done.set()

    client = create_client(topics)
    client.on_message = on_message
    client.connect_async("price-streaming-free.vndirect.com.vn", 443, keepalive=30)
    loop_thread = threading.Thread(target=client.loop_forever, daemon=True)
    loop_thread.start()

    if not client._userdata["connected"].wait(15):
        raise RuntimeError(
            "Khong ket noi duoc MQTT realtime cua VNDIRECT trong 15 giay."
        )

    done.wait(WAIT_TIMEOUT_SECONDS)
    client.disconnect()
    loop_thread.join(timeout=5)
    return sp_map, ba_map


def main():
    all_codes = load_stock_codes()
    selected_codes = select_codes(all_codes, sys.argv[1:])
    fetched_at = datetime.now().strftime("%Y-%m-%d %H:%M:%S")
    sp_map, ba_map = collect_realtime_rows(selected_codes)

    if not sp_map and not ba_map:
        raise SystemExit(
            "Khong nhan duoc ban tin realtime nao. Co the dang ngoai gio giao dich "
            "hoac feed free khong phat sinh du lieu cho cac ma vua chon."
        )

    normalized_rows = [
        merge_stock_data(code, sp_map.get(code), ba_map.get(code), fetched_at)
        for code in selected_codes
    ]
    content = "\n\n".join(
        format_block(index, record)
        for index, record in enumerate(normalized_rows, start=1)
    )
    OUTPUT_FILE.write_text(content + "\n", encoding="utf-8")

    print("Da chon 5 ma co phieu:")
    for code in selected_codes:
        print(f"- {code}")
    print(f"\nDa luu ket qua vao: {OUTPUT_FILE}")


if __name__ == "__main__":
    main()

File: generate_bang_gia.R
#!/usr/bin/env Rscript

if (!requireNamespace("jsonlite", quietly = TRUE)) {
  stop(
    paste(
      "Missing dependency: jsonlite.",
      "Install it with: install.packages('jsonlite')"
    ),
    call. = FALSE
  )
}

base_dir <- normalizePath(dirname(commandArgs(trailingOnly = FALSE)[grep("^--file=", commandArgs(trailingOnly = FALSE))]), winslash = "/", mustWork = TRUE)
stock_id_dir <- file.path(base_dir, "StockIDs")
output_file <- file.path(base_dir, "bang_gia_chung_khoan.txt")
random_count <- 5L
snapshot_api_url <- "https://price-streaming-api-free.vndirect.com.vn/v2/stocks/snapshot"

pick_value <- function(..., default = "N/A") {
  values <- list(...)
  for (value in values) {
    if (!is.null(value) && !identical(value, "") && !is.na(value)) {
      return(value)
    }
  }
  default
}

load_stock_codes <- function() {
  files <- sort(list.files(stock_id_dir, pattern = "\\.txt$", full.names = TRUE))
  codes <- character()

  for (path in files) {
    lines <- readLines(path, warn = FALSE, encoding = "UTF-8")
    cleaned <- toupper(trimws(lines))
    cleaned <- cleaned[nzchar(cleaned)]
    codes <- c(codes, cleaned)
  }

  unique_codes <- sort(unique(codes))
  if (length(unique_codes) < random_count) {
    stop(sprintf("Not enough stock codes to select %d random symbols.", random_count), call. = FALSE)
  }

  unique_codes
}

select_codes <- function(all_codes, args) {
  if (length(args) == 0L) {
    return(sample(all_codes, random_count))
  }

  selected_codes <- toupper(trimws(args))
  selected_codes <- selected_codes[nzchar(selected_codes)]

  if (length(selected_codes) != random_count) {
    stop(sprintf("Please provide exactly %d stock symbols.", random_count), call. = FALSE)
  }

  invalid_codes <- selected_codes[!selected_codes %in% all_codes]
  if (length(invalid_codes) > 0L) {
    stop(sprintf("Invalid symbols: %s", paste(invalid_codes, collapse = ", ")), call. = FALSE)
  }

  if (length(unique(selected_codes)) != random_count) {
    stop("All 5 stock symbols must be different.", call. = FALSE)
  }

  selected_codes
}

decode_message <- function(encoded_message) {
  chars <- strsplit(encoded_message, "", fixed = TRUE)[[1]]
  decoded <- mapply(
    function(char, index) intToUtf8(utf8ToInt(char) + ((index - 1L) %% 5L)),
    chars,
    seq_along(chars),
    USE.NAMES = FALSE
  )
  paste(decoded, collapse = "")
}

decode_fields <- function(encoded_message) {
  strsplit(decode_message(encoded_message), "|", fixed = TRUE)[[1]]
}

map_fields <- function(field_names, values) {
  result <- as.list(rep(NA_character_, length(field_names)))
  names(result) <- field_names
  count <- min(length(field_names), length(values))
  result[seq_len(count)] <- as.list(values[seq_len(count)])
  result
}

parse_sfu <- function(values) {
  stock_type <- values[[2]]

  if (identical(stock_type, "ST")) {
    fields <- c(
      "code", "stockType", "floorCode", "basicPrice", "floorPrice", "ceilingPrice",
      "bidPrice01", "bidPrice02", "bidPrice03", "bidPrice04", "bidPrice05",
      "bidPrice06", "bidPrice07", "bidPrice08", "bidPrice09", "bidPrice10",
      "bidQtty01", "bidQtty02", "bidQtty03", "bidQtty04", "bidQtty05",
      "bidQtty06", "bidQtty07", "bidQtty08", "bidQtty09", "bidQtty10",
      "offerPrice01", "offerPrice02", "offerPrice03", "offerPrice04", "offerPrice05",
      "offerPrice06", "offerPrice07", "offerPrice08", "offerPrice09", "offerPrice10",
      "offerQtty01", "offerQtty02", "offerQtty03", "offerQtty04", "offerQtty05",
      "offerQtty06", "offerQtty07", "offerQtty08", "offerQtty09", "offerQtty10",
      "totalBidQtty", "totalOfferQtty", "highestPrice", "lowestPrice",
      "accumulatedVal", "accumulatedVol", "matchPrice", "matchQtty",
      "currentPrice", "currentQtty", "totalRoom", "currentRoom", "iNav",
      "underlyingAsset", "issuer", "exercisePrice", "exerciseRatio",
      "expiryDate", "time", "bv4", "sv4"
    )
    return(map_fields(fields, values))
  }

  fields <- c(
    "code", "stockType", "floorCode", "basicPrice", "floorPrice", "ceilingPrice",
    "bidPrice01", "bidPrice02", "bidPrice03", "bidQtty01", "bidQtty02", "bidQtty03",
    "offerPrice01", "offerPrice02", "offerPrice03", "offerQtty01", "offerQtty02",
    "offerQtty03", "totalBidQtty", "totalOfferQtty", "tradingSessionId",
    "buyForeignQtty", "sellForeignQtty", "highestPrice", "lowestPrice",
    "accumulatedVal", "accumulatedVol", "matchPrice", "matchQtty", "currentPrice",
    "currentQtty", "projectOpen", "totalRoom", "currentRoom", "iNav"
  )
  map_fields(fields, values)
}

fetch_snapshot_rows <- function(selected_codes) {
  query <- paste(utils::URLencode(selected_codes, reserved = TRUE), collapse = ",")
  url <- sprintf("%s?codes=%s", snapshot_api_url, query)
  raw_rows <- jsonlite::fromJSON(url)

  sp_map <- list()
  for (raw_row in raw_rows) {
    fields <- decode_fields(raw_row)
    if (length(fields) == 0L) {
      next
    }

    message_type <- fields[[1]]
    body <- fields[-1]

    if (identical(message_type, "SFU")) {
      data <- parse_sfu(body)
      code <- data$code
      if (!is.null(code) && code %in% selected_codes) {
        sp_map[[code]] <- data
      }
    }
  }

  sp_map
}

merge_stock_data <- function(code, sp_data, fetched_at) {
  list(
    "Mã chứng khoán" = code,
    "Giá tham chiếu" = pick_value(sp_data$basicPrice),
    "Giá trần" = pick_value(sp_data$ceilingPrice),
    "Giá sàn" = pick_value(sp_data$floorPrice),
    "Giá mua 1" = pick_value(sp_data$bidPrice01),
    "Khối lượng mua 1" = pick_value(sp_data$bidQtty01),
    "Giá mua 2" = pick_value(sp_data$bidPrice02),
    "Khối lượng mua 2" = pick_value(sp_data$bidQtty02),
    "Giá mua 3" = pick_value(sp_data$bidPrice03),
    "Khối lượng mua 3" = pick_value(sp_data$bidQtty03),
    "Giá bán 1" = pick_value(sp_data$offerPrice01),
    "Khối lượng bán 1" = pick_value(sp_data$offerQtty01),
    "Giá bán 2" = pick_value(sp_data$offerPrice02),
    "Khối lượng bán 2" = pick_value(sp_data$offerQtty02),
    "Giá bán 3" = pick_value(sp_data$offerPrice03),
    "Khối lượng bán 3" = pick_value(sp_data$offerQtty03),
    "Giá khớp lệnh" = pick_value(sp_data$matchPrice),
    "Khối lượng khớp lệnh" = pick_value(sp_data$matchQtty),
    "Thời gian lấy dữ liệu" = pick_value(sp_data$time, fetched_at)
  )
}

format_block <- function(index, record) {
  lines <- c(sprintf("Co phieu %d", index))
  for (name in names(record)) {
    lines <- c(lines, sprintf("%s: %s", name, record[[name]]))
  }
  paste(lines, collapse = "\n")
}

main <- function() {
  all_codes <- load_stock_codes()
  selected_codes <- select_codes(all_codes, commandArgs(trailingOnly = TRUE))
  fetched_at <- format(Sys.time(), "%Y-%m-%d %H:%M:%S")
  sp_map <- fetch_snapshot_rows(selected_codes)

  rows <- lapply(
    selected_codes,
    function(code) merge_stock_data(code, sp_map[[code]], fetched_at)
  )

  content <- paste(
    vapply(seq_along(rows), function(index) format_block(index, rows[[index]]), character(1)),
    collapse = "\n\n"
  )
  writeLines(c(content, ""), output_file, useBytes = TRUE)

  cat("Da chon 5 ma co phieu:\n")
  for (code in selected_codes) {
    cat(sprintf("- %s\n", code))
  }
  cat(sprintf("\nDa luu ket qua vao: %s\n", output_file))
}

main()
